src/UCTSearch.cpp

```diff
@@ -49,12 +50,71 @@ void UCTSearch::update_root() {
     // Definition of m_playouts is playouts per search call.
     // So reset this count now.
     m_playouts = 0;
-    // See if the current state of the game is a direct succesor of
-    // our stored tree's root node. If so, make that the new root.
-    UCTNode::find_new_root(m_root, m_rootstate, std::move(m_last_rootstate));
-    // If the above succeeded, we'll have a search tree already. See
-    // how big it is.
+
+    bool found = false;
+    #ifndef NDEBUG
+    auto start_nodes = m_root->count_nodes();
+    std::string replay;
+    #endif
+
+    // Check if new rootstate is a possible descendant of our stored rootstate.
+    if (m_last_rootstate
+            && m_rootstate.get_komi() == m_last_rootstate->get_komi()
+            && m_rootstate.get_movenum() >= m_last_rootstate->get_movenum()) {
+        auto depth = m_rootstate.get_movenum() - m_last_rootstate->get_movenum();
+        auto test = std::make_unique<GameState>(m_rootstate);
+        for (auto i = size_t{0}; i < depth; i++) {
+            test->undo_move();
+        }
+
+        // Try to replay moves advancing m_root
+        for (auto i = size_t{0}; i < depth; i++) {
+            // Verify m_rootstate and m_last_rootstate have same history.
+            if (m_last_rootstate->board.get_ko_hash()
+                    != test->board.get_ko_hash()) {
+                break;
+            }
+
+            test->forward_move();
+            auto move = test->get_last_move();
+            m_root = m_root->find_child(move);
+            if (!m_root->has_children()) {
+                break;
+            }
+            m_last_rootstate->play_move(move);
+            # ifndef NDEBUG
+            replay += m_last_rootstate->board.move_to_text(move) + " ";
+            # endif
+        }
+
+        // Check if advancing last_rootstate was possible.
+        if (m_root->has_children()
```

Comment:
I think this separate check can go away. The code that finds the match and sets the new root can set found. If it did not, nothing was found.

Reply:
I think I would lose the last ko_hash check if I do it that way + I have to set found = true at the start of the if which isn't as pretty to me.

Reply:
I thought about this some more.

I believe I only need the first ko hash check, if a position has the same ko_hash then you play a series of identical moves they will have the same end ko_hash.

But this made me realize that this potentially returns trees which are transposes of the real position.

if you
play Q16, D4, R4, D16
undo x4
play R4, D4, Q16, D16
it will reuse the nodes (but they will have been generate with the moves in a different order).

This isn't an issue for the autogtp, or any of the bots but for humans doing self play it can technically happen.

Reply:
Yes, that would be undesirable.

Reply:
I found a case where this can fail.

If the user has entered two black moves in a row the node will find the child but the hashes won't match :)